src/monitors/R3Monitor.v3

```diff
@@ -0,0 +1,572 @@
+				var export_func_name = exported_func_names[index];
+				trace.put(WasmEvent.FuncEntry(index, export_func_name, params));
+			} else { // called with table_get
+				var index_match_found = false;
+				for (table in instance.tables) {
+					for (table_idx < table.elems.length) {
+						if (index == resolveFuncIdx(table, table_idx)) {
+							trace.put(WasmEvent.FuncEntryTable(index, table_idx, params));
+							index_match_found = true;
+							break;
+						}
+					}
+					if (index_match_found) break;
+				}
+				System.error("R3MonitorError", "external call with table_get failed");
+			}
+			checkMemGrow(instance);
+			checkTableGrow(instance);
+		}
+		call_kind_stack.push((CallKind.INT, index));
+	}
+
+	def onFuncExit(index: int) {
+		call_kind_stack.pop();
+	}
+
+	def onCall(func_id: int) {
+		var func_decl = module.functions[func_id];
+		if (func_decl.imp != null) {
+			last_instr_is_call = true;
+			call_kind_stack.push((CallKind.EXT, func_id));
+			trace.put(WasmEvent.ImportCall(func_id));
+		}
+	}
+
+	def onIndirectCall(table_index: int, entry_idx: int, table: Table) {
+		var func_id = onTableGet(table_index, entry_idx, table);
+		if (func_id != -1) {
+			onCall(func_id);
+		}
+	}
+
+	def onCallReturn(loc: DynamicLoc) {
+		if (call_kind_stack.peek().0 == CallKind.INT) {
+			return;
+		}
+		var frame_accessor = loc.frame.getFrameAccessor();
+		var func_id = call_kind_stack.pop().1;
+		var func = module.functions[func_id];
+		var results_len = func.sig.results.length;
+		var results = Array<Value>.new(results_len);
+		var func_name = getFuncName(func.func_index);
+		for (i < results_len) {
+			results[i] = frame_accessor.getOperand(-1 * (results_len - i - 1));
+		}
+		trace.put(WasmEvent.ImportReturn(results));
+		var instance = loc.func.instance;
+		checkMemGrow(instance);
+		checkTableGrow(instance);
+	}
+
+	def onGlobalSet(id: int, new_value: Value) {
+		shadowGlobals[id] = new_value;
+	}
+
+	def onGlobalGet(id: int, value: Value) {
+		if (shadowGlobals[id] != value) {
+			trace.put(WasmEvent.GlobalGet(id, value));
+			shadowGlobals[id] = value;
+		}
+	}
+
+	def onTableGet(table_idx: int, entry_idx: int, table: Table) -> int {
+		var func_idx = resolveFuncIdx(table, entry_idx);
+		if (func_idx != -1) {
+			var func_name = getFuncName(func_idx);
+			if (table.elems[entry_idx] != shadowTables[table_idx].elems[entry_idx]) {
+				trace.put(WasmEvent.TableGet(table_idx, entry_idx, func_idx, func_name));
+				shadowTables[table_idx][entry_idx] = table.elems[entry_idx];
+			}
+		}
+		return func_idx;
+	}
+
+	def onTableSet(table_idx: int, entry_idx: int, new_val: Value) {
+		shadowTables[table_idx][entry_idx] = new_val;
+	}
+
+	def onTableGrow(table_idx: int, growth: u32) {
+		trace.put(WasmEvent.TableGrow(table_idx, growth));
+		shadowTables[table_idx].grow(growth, Values.FUNCREF_NULL);
+	}
+
+	def memoryRead(loc: DynamicLoc, mem: Memory, addr: u64, size: u64) -> Resumption {
+		var addr = u32.view(addr);
```

Comment:
If you use the 64-bit memory range methods, you won't need to do this view, and it will be forward-compatible with Wasm memory64.